### DISCUSSION TITLE
Sentry integration

### DISCUSSION THREAD

**Discussion by soukicz** [brefphp/bref#966] [Support]
Can you please help me integrate Sentry? I have tried to just add `\Sentry\init(['dsn' => $_ENV['SENTRY_DSN']]);` to handler, but that doesn't catch anything. My guess is that bref is already doing error handling and I would have to inject Sentry somewhere to send unhandled exceptions manually.

I have tried to search documentation but there is only mention (https://bref.sh/docs/environment/logs.html#php-errors-and-warnings) that errors are written to logs but no information on how to extend it.

Can you please point me to right direction? Thank you

**Comment by soukicz** [comment]
I am currently just wrapping handler in this code but it doesn't feel right:

```php
public function handle($event, BrefContext $context) {
   try {

         // handler body

   } catch (\Throwable $exception) {
          \Sentry\captureException($exception);
          throw $exception;
   }
}
```

**Reply by mnapoli** [reply]
Hi, that sounds like a reasonable approach. If you want to capture exceptions in your code, then you have to try/catch it.

It's good that you re-throw the exception too because that will mark the Lambda invocation as failed. This can be important for metrics, any CloudWatch alert you may setup, SQS queue processing, etc.

TL/DR: looks good!

**Reply by soukicz** [reply]
Thank you!

I didn't like that I have to wrap each handler in try/catch but after good night sleep I realized, that I could just move this to shared handler class.

**Reply by mnapoli** [reply]
Yes, are you writing an API? Or anything else?

**Reply by t-richard** [reply]
FYI this is more or less what is done by the Sentry JS SDK for Lambda

```js
exports.handler = Sentry.AWSLambda.wrapHandler(async (event, context) => {
  // Your handler code
});
```

It basically wraps the handler in a try catch and does some more things to enhance reporting by adding metadata about the execution context of the Lambda.

**Reply by soukicz** [reply]
@t-richard  That is exactly where I copied it from :)) I was worried that it wouldn't catch fatal errors etc but that isn't problem with PHP8. I can't really think of another way that would crash the app and wouldn't be caught by Sentry.

**Reply by mnapoli** [reply]
Ah Sentry has a thing that can catch fatal errors, it needs to be registered at the top of the script and that should be good.

Wrapping the handler function is a good idea!

**Reply by soukicz** [reply]
Registering Sentry without try/catch didn't work for me. I thought that either Bref catches the error first or Sentry just doesn't expects this kind of request life cycle.

**Reply by mnapoli** [reply]
Registering will work for fatal errors only, that's what I meant.

I would do:

- register (for errors/warnings, i.e. non-exceptions)
- try/catch (for exceptions)
